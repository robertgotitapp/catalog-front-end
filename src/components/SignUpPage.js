import React, { Component } from 'react';
import { connect } from 'react-redux';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';
import { Redirect } from 'react-router-dom';
import { signUp, signIn, getUserData } from '../actions/users';
import { REGEX } from '../utils/const';

export class SignUpPage extends Component {
    state = {
      username: '',
      password: '',
      name: '',
      email: '',
      toHome: false,
      toSignIn: false,
      alerts: '',
    }

    validateInput = () => {
      const {
        username, password, name, email,
      } = this.state;
      const errors = {};
      if (username.length <= 5) {
        errors.username = 'Username must be at least 5 characters.';
      }
      if (password.length < 8) {
        errors.password = 'Password must be at least 8 characters';
      }
      if (name.length <= 5) {
        errors.name = 'Name must be at least 5 characters.';
      }
      if (email.length < 8) {
        errors.email = 'Email must be at least 8 characters';
      }
      console.log(errors);
      if (Object.keys(errors).length !== 0) {
        this.setState({
          alerts: errors,
        });
        return false;
      }
      return true;
    }

    handleSubmit = (e) => {
      e.preventDefault();
      const {
        username, password, name, email,
      } = this.state;
      if (this.validateInput()) {
        this.props.signUp({
          username, password, name, email,
        })
          .then((res) => {
            if (res.statusCode) {
              // If request is successful, proceed to automatically sign in
              this.props.signIn({ username, password })
                .then((secondRes) => {
                  // If sign in is successful, relocate to Home page,
                  // Otherwise, go to SignIn page
                  if (secondRes.statusCode) {
                    this.props.getUserData()
                      .then((thirdRes) => {
                        if (thirdRes.statusCode) {
                          this.setState(prevState => ({ ...prevState, toHome: true }));
                        }
                      });
                  } else {
                    this.setState(prevState => ({ ...prevState, toSignIn: true }));
                  }
                });
            } else {
              // If sign up request is failed, reset all the form input
              // and display all the error messages as alerts
              res.errorPromise
                .then((error) => {
                  this.setState({
                    username: '',
                    password: '',
                    name: '',
                    email: '',
                    toHome: false,
                    toSignIn: false,
                    alerts: error.message,
                  });
                });
            }
          });
      }
    }

    handleChange = (e) => {
      this.setState({
        [e.target.name]: e.target.value,
      });
    }

    render() {
      const {
        username, password, name, email, toHome, toSignIn, validationErrors,
      } = this.state;

      if (toHome) {
        return <Redirect to='/' />;
      }
      if (toSignIn) {
        return <Redirect to='/signin' />;
      }

      return (
        <div>
          <h2>Sign Up</h2>
          <Form onSubmit={this.handleSubmit}>
            <Form.Group>
              <Form.Label>Username</Form.Label>
              <Form.Control
                type="text"
                name="username"
                onChange={this.handleChange}
                value={username}
              />
            </Form.Group>
            <Form.Group>
              <Form.Label>Password</Form.Label>
              <Form.Control type="password" name="password" onChange={this.handleChange} value={password} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Email</Form.Label>
              <Form.Control type="text" name="email" onChange={this.handleChange} value={email} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Name</Form.Label>
              <Form.Control type="text" name="name" onChange={this.handleChange} value={name} />
            </Form.Group>
            <Button variant="primary" type="submit">
              Submit
            </Button>
          </Form>
          {
            // Loop through alerts dictionary and display an alert for each
            // of one item in dictionary
            this.state.alerts
            && Object.keys(this.state.alerts).map((key, id) => (
              <Alert key={id} variant='danger'>
                {key}
                :
                {this.state.alerts[key]}
              </Alert>
            ))
          }
        </div>
      );
    }
}

const mapDispatchToProps = {
  signUp, signIn, getUserData,
};

export default connect(null, mapDispatchToProps)(SignUpPage);
