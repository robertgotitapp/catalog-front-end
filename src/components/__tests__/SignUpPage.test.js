import React from 'react';
import { shallow } from 'enzyme';
import Form from 'react-bootstrap/Form';
import Alert from 'react-bootstrap/Alert';
import { SignUpPage } from '../SignUpPage';

describe('components/SignUpPage', () => {
  const props = {};
  const container = shallow(<SignUpPage {...props} />);

  it('should render correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('should contain 1 FormControl for username', () => {
    expect(container.find('[name="username"]')).toHaveLength(1);
  });

  it('should contain 1 FormControl for password', () => {
    expect(container.find('[name="password"]')).toHaveLength(1);
  });

  it('should contain 1 FormControl for email', () => {
    expect(container.find('[name="email"]')).toHaveLength(1);
  });

  it('should contain 1 FormControl for name', () => {
    expect(container.find('[name="name"]')).toHaveLength(1);
  });

  it('should trigger state change when FormControl username got input', () => {
    const usernameInput = container.find('[name="username"]');
    usernameInput.simulate('change', { target: { name: 'username', value: 'user1' } });
    expect(container.state().username).toEqual('user1');
  });

  it('should trigger state change when FormControl password got input', () => {
    const passwordInput = container.find('[name="password"]');
    passwordInput.simulate('change', { target: { name: 'password', value: 'Password1' } });
    expect(container.state().password).toEqual('Password1');
  });

  it('should trigger state change when FormControl email got input', () => {
    const emailInput = container.find('[name="email"]');
    emailInput.simulate('change', { target: { name: 'email', value: 'user1@gmail.com' } });
    expect(container.state().email).toEqual('user1@gmail.com');
  });

  it('should trigger state change when FormControl name got input', () => {
    const nameInput = container.find('[name="name"]');
    nameInput.simulate('change', { target: { name: 'name', value: 'User Name' } });
    expect(container.state().name).toEqual('User Name');
  });

  it('should trigger handler when form is submited', () => {
    container.instance().handleSubmit = jest.fn();
    container.instance().forceUpdate();
    const form = container.find(Form);
    form.simulate('submit');
    expect(container.instance().handleSubmit).toHaveBeenCalled();
  });

  it('should trigger alert when there is error message', () => {
    container.setState({
      username: '',
      password: '',
      name: '',
      email: '',
      toHome: false,
      toSignIn: false,
      alerts: {
        email: 'has been registed with different account',
      },
    });
    expect(container.find(Alert)).toHaveLength(1);
  });
});
