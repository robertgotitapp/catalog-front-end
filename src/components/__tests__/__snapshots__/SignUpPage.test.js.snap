// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SignUpPage should render correctly 1`] = `
<div>
  <Form
    inline={false}
    onSubmit={[Function]}
  >
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Username
      </FormLabel>
      <FormControl
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Password
      </FormLabel>
      <FormControl
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Email
      </FormLabel>
      <FormControl
        name="email"
        onChange={[Function]}
        type="text"
        value=""
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Name
      </FormLabel>
      <FormControl
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </FormGroup>
    <Button
      active={false}
      disabled={false}
      type="submit"
      variant="primary"
    >
      Submit
    </Button>
  </Form>
</div>
`;
